File: Exavision.Seasense.Shared/Capabilities/Turret/TurretMoveSpeedCapability.cs
using Exavision.Seasense.Shared.Settings;

namespace Exavision.Seasense.Shared.Capabilities.Turret {
    public abstract class TurretMoveSpeedCapability<S> : Capability<S>, ITurretMoveSpeedCapability where S : SettingCapability, new() {




        public override S GetSetting(S setting) {
            setting.CapabilityType = CapabilityType.TurretMoveSpeed;
            return setting;
        }

        public abstract void MoveSpeed(double axisX, double axisY);
    }
}

File: Exavision.Seasense/Client/src/app/materials/capabilities/turret/turret-move-speed-capability.ts
import { Material } from "../../material";
import { SettingCapability } from "../../settings/setting-capability";
import { Site } from "../../site";
import { CapabilityState } from "../../states/capability-state";
import { Unit } from "../../unit";
import { Capability } from "../capability";
import { CapabilityType } from "../capability-type";

export class TurretMoveSpeedCapability implements Capability {
 
  id: string;
  capabilityType: CapabilityType;
  material: Material | undefined;


  constructor(settingCapability: SettingCapability) {
    this.id = settingCapability.id;
    this.capabilityType = CapabilityType.TurretMoveSpeed;
  }

  moveSpeed(axisX: number, axisY: number) {
    if (this.material == null) return;
    if (this.material.unit == null) return;
    this.material?.wsService.turretMoveSpeed(this.material.unit.id, this.material.id, axisX, axisY);
    
  }
  setState(valueState: CapabilityState): void {

  }
}
